components/kyma-environment-broker/internal/process/deprovisioning/btp_operator_cleanup.go

```diff
@@ -0,0 +1,135 @@
+	if !operation.Temporary {
+		log.Info("cleanup executed only for suspensions")
+		return operation, 0, nil
+	}
+	if operation.RuntimeID == "" {
+		log.Info("instance has been deprovisioned already")
+		return operation, 0, nil
+	}
+	status, err := s.provisionerClient.RuntimeStatus(operation.ProvisioningParameters.ErsContext.GlobalAccountID, operation.RuntimeID)
+	if err != nil {
+		if s.isNotFoundErr(err) {
+			log.Info("instance not found in provisioner")
+			return operation, 0, nil
+		}
+		return handleError(s.Name(), operation, err, log, "call to provisioner RuntimeStatus failed")
+	}
+	if status.RuntimeConfiguration.Kubeconfig == nil {
+		err := kebError.NewTemporaryError("empty kubeconfig")
+		return handleError(s.Name(), operation, err, log, "provisioner returned empty kubeconfig")
+	}
+	k := *status.RuntimeConfiguration.Kubeconfig
+	hash := sha256.Sum256([]byte(k))
+	log.Infof("kubeconfig details length: %v, sha256: %v", len(k), string(hash[:]))
+	if len(k) < 10 {
+		err := kebError.NewTemporaryError("kubeconfig suspiciously small, requeuing")
+		return handleError(s.Name(), operation, err, log, "provisioner returned wrong kubeconfig")
+	}
+	cli, err := s.k8sClientProvider(k)
+	if err != nil {
+		err = kebError.AsTemporaryError(err, "failed to create k8s client from the kubeconfig")
+		return handleError(s.Name(), operation, err, log, "could not create a k8s client")
+	}
+	if err := s.deleteServiceBindingsAndInstances(cli, log); err != nil {
+		err = kebError.AsTemporaryError(err, "failed BTP operator resource cleanup")
+		return handleError(s.Name(), operation, err, log, "could not delete bindings and service instances")
+	}
+	return operation, 0, nil
+}
+
+func (s *BTPOperatorCleanupStep) deleteServiceBindingsAndInstances(k8sClient client.Client, log logrus.FieldLogger) error {
+	namespaces := corev1.NamespaceList{}
+	if err := k8sClient.List(context.Background(), &namespaces); err != nil {
+		return err
+	}
+	requeue := s.deleteResource(k8sClient, namespaces, schema.GroupVersionKind{Group: btpOperatorGroup, Version: btpOperatorApiVer, Kind: btpOperatorBinding}, log)
+	requeue = requeue || s.deleteResource(k8sClient, namespaces, schema.GroupVersionKind{Group: btpOperatorGroup, Version: btpOperatorApiVer, Kind: btpOperatorServiceInstance}, log)
+	if requeue {
+		return fmt.Errorf("waiting for resources to be deleted")
+	}
+	return nil
+}
+
+func (s *BTPOperatorCleanupStep) deleteResource(k8sClient client.Client, namespaces corev1.NamespaceList, gvk schema.GroupVersionKind, log logrus.FieldLogger) (requeue bool) {
+	listGvk := gvk
+	listGvk.Kind = gvk.Kind + "List"
+	stillExistingCount := 0
+	for _, ns := range namespaces.Items {
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(listGvk)
+		if err := k8sClient.List(context.Background(), list, client.InNamespace(ns.Name)); err != nil {
+			log.Errorf("failed listing resource %v in namespace %v", gvk, ns.Name)
```

Comment:
the deletion follows listing because it's desirable to allow BTP-Operator to properly delete these objects and perform finalization logic. The cleanup logic follows these steps:

1) list `ServiceInstances` and `ServiceBindings` in all namespaces
2) if none exist, consider cleanup as finished
3) if some do exist, call `DeleteAll` and requeue after 10s (limit up to 30 minutes)
4) the operation may be requeued a few times but eventually terminate as BTP-Operator removes all relevant resources